### PR TITLE
[WIP] Page.get_absolute_url should actually return a site domain aware version of the page's url

### PR DESCRIPTION
**Pull request by heppstux**
see issue #5610

### CODE REVIEW

**Comment by czpython** [comment]
Hello @heppstux ,
Thanks for the patch.
I'd rather we have a different method on the `Page` model that always return a full url w/ domain and all, instead of overloading the `get_absolute_url()` method.

This is for several reasons:

- Backwards compatibility
- Consistency with Django
- Consistency in behavior (always return full url vs only when it doesn't match current site)

We can call the new method `get_full_url()`, takes same parameters as `get_absolute_url()` but returns the url with domain.

Then we can adapt only the parts in the code that need this behavior like the link plugin, etc..

**Comment by heppstux** [comment]
okay, sounds reasonable – however a lot of (third-party) plugins will not benefit automatically from this adjustments. Also I feel, it's the actual job of the get_absolute_url to actually do what it's name suggests and return the real url of a Page.

How can we make sure, all parts of the CMS benefit from this new function other than deprecate the get_absolute_url method...

**Comment by yakky** [comment]
@heppstux Right or wrong ``get_absolute_url`` only return the **path** of a model, not the full URL.
I don't think django CMS should break this in anyway and I'd go for an additional method as suggested by Paulo, without deprecating ``get_absolute_url``

**Comment by heppstux** [comment]
Okay – I can live with that.

Even though I see some more problems and even more confusion coming: Should get_draft_url and get_public_url also get alternative "versions" with explicit support for Sites?

Let's say "get_full_draft_url" and "get_full_public_url".

**Comment by yakky** [comment]
get_draft_url and get_public_url are only meant to be used within the frontend editor, I don't think they  need a full url version

**Comment by czpython** [comment]
I agree, the draft / public methods should only be used internally by the cms and as such, is not needed to create a "full" version for them.

File: cms/models/pagemodel.py [hunk 1]
```diff
@@ -174,10 +175,15 @@ def get_absolute_url(self, language=None, fallback=True):
...
         with i18n.force_language(language):
-            if self.is_home:
-                return reverse('pages-root')
             path = self.get_path(language, fallback) or self.get_slug(language, fallback)
-            return reverse('pages-details-by-slug', kwargs={"slug": path})
+            if self.is_home:
+                relative_url = reverse('pages-root')
+            else:
+                relative_url = reverse('pages-details-by-slug', kwargs={"slug": path})
+            if self.site:
```
Comment (vstoykov) [inline cms/models/pagemodel.py]: This check can be optimized to reduce SQL queries when page's site is the same as configured in settings.

    if self.site_id and settings.SITE_ID != self.site_id and self.site.domain:
        return '//%s%s' % (self.site.domain, relative_url)

Actually I'm not sure if the check for site.domain is needed at all. `domain` field is not only required but also unique and if there is a `site_id` then there will be a `site.domain`.

Comment (heppstux) [inline cms/models/pagemodel.py]: yes, thanks :)

**Comment by czpython** [comment]
Hello @heppstux, any updates?

**Comment by heppstux** [comment]
hi @czpython, sorry, didn't have the chance to do any further investigation/coding on the topic yet – still it remains hot. I'd suggest to reject this PR and I'll try to open another one later.

**Comment by czpython** [comment]
Thanks for the follow-up.
Closing as suggested.

**Pull request closed without merging**